src/main/native/pdhmm/pdhmm-serial.cc

```diff
@@ -0,0 +1,462 @@
+
+			// Special case since we MIGHT be at the deletion end and have to be handling jump states from above
+			if ((hap_pdbases[j - OFF] & PartiallyDeterminedHaplotype::DEL_END) == PartiallyDeterminedHaplotype::DEL_END)
+			{
+				insertionMatrix[j] = std::max(bmmTop, mmTop) * currTransition[matchToInsertion] + std::max(bimTop, imTop) * currTransition[insertionToInsertion];
+			}
+			else
+			{
+				insertionMatrix[j] = mmTop * currTransition[matchToInsertion] + imTop * currTransition[insertionToInsertion];
+			}
+
+			if (currentState == AFTER_DEL)
+			{
+				currentState = NORMAL;
+			}
+
+			// If we are at a deletion start base, start copying the branch states
+			if ((hap_pdbases[j - 1] & PartiallyDeterminedHaplotype::DEL_START) == PartiallyDeterminedHaplotype::DEL_START)
+			{
+				currentState = INSIDE_DEL;
+			}
+			// Being after a deltion overrides being inside a deletion by virtue of the fact that we allow single element deletions
+			if ((hap_pdbases[j - 1] & PartiallyDeterminedHaplotype::DEL_END) == PartiallyDeterminedHaplotype::DEL_END)
+			{
+				currentState = AFTER_DEL;
+			}
+
+			bmmDiag = bmmTop;
+			bimDiag = bimTop;
+			bdmDiag = bdmTop;
+			mmDiag = mmTop;
+			imDiag = imTop;
+			dmDiag = dmTop;
+			bmmLeft = branchMatchMatrix[j];
+			bimLeft = branchInsertionMatrix[j];
+			bdmLeft = branchDeletionMatrix[j];
+			mmLeft = matchMatrix[j];
+			imLeft = insertionMatrix[j];
+			dmLeft = deletionMatrix[j];
+		}
+	}
+
+	// final log probability is the log10 sum of the last element in the Match and Insertion state arrays
+	// this way we ignore all paths that ended in deletions! (huge)
+	// but we have to sum all the paths ending in the M and I matrices, because they're no longer extended.
+	int32_t endI = paddedReadLength - 1;
+	double finalSumProbabilities = 0.0;
+	for (int32_t j = 1; j < paddedHaplotypeLength; j++)
+	{
+		finalSumProbabilities += matchMatrix[j] + insertionMatrix[j];
+	}
+	return log10(finalSumProbabilities) - INITIAL_CONDITION_LOG10;
+}
+
+double computeReadLikelihoodGivenHaplotypeLog10(const int8_t *hap_bases, const int8_t *hap_pdbases, const int8_t *read_bases, const int8_t *read_qual, const int8_t *read_ins_qual, const int8_t *read_del_qual, const int8_t *gcp, int64_t hap_bases_length, int64_t read_bases_length, int32_t &prev_hap_bases_length, double *matchMatrix, double *insertionMatrix, double *deletionMatrix, double *branchMatchMatrix, double *branchInsertionMatrix, double *branchDeletionMatrix, bool &constantsAreInitialized, double *transition, double *prior, const double *matchToMatchProb, const double *qualToErrorProbCache, int32_t &status, int32_t maxHaplotypeLength)
+{
+	status = PDHMM_SUCCESS;
+	bool recacheReadValues = false;
+	int32_t hapStartIndex = 0;
+
+	int32_t paddedReadLength = (int32_t)read_bases_length + 1;
+	int32_t paddedHaplotypeLength = (int32_t)hap_bases_length + 1;
+
+	hapStartIndex = (recacheReadValues) ? 0 : hapStartIndex;
```

Comment:
This is always false

Reply:
I agree. However, this is intentionally maintained to ensure that we do not execute fewer instructions compared to PDHMM in GATK, and any performance gain is solely attributed to our optimizations.